File: src/olcw78 (die yoong)/jsAlgorithmStudy/week02/06_실패율.go
package main

import (
	"fmt"
	"slices"
	"sort"
)

type Res struct {
	i        int
	failRate float64
}

func solution(N int, stages []int) []int {
	fails := make([]Res, 0, N)

	for i := 1; i <= N; i++ {
		yet := 0
		reached := 0

		del := make([]int, 0)
		for sti, st := range stages {
			reached++
			if st <= i {
				yet++
				del = append(del, sti)
			}
		}

		deletedStages := make([]int, len(stages), len(stages))
		copy(deletedStages, stages)

		sort.Sort(sort.Reverse(sort.IntSlice(del)))
		for _, e := range del {
			deletedStages = append(deletedStages[:e], deletedStages[e+1:]...)
		}

		stages = stages[:0]
		stages = append(stages, deletedStages...)

		failRate := float64(yet) / float64(reached)
		fails = append(fails, Res{i, failRate})
	}

	slices.SortFunc(
		fails,
		func(a, b Res) int {
			v := b.failRate - a.failRate
			v *= 10000
			return int(v)
		})

	res := make([]int, 0, len(fails))
	for _, fail := range fails {
		res = append(res, fail.i)
	}

	return res
}

type Res2 struct {
	stage    int
	failRate float64
}

func solution1(N int, stages []int) []int {
	failRates := make([]Res2, N)
	for stage := 1; stage <= N; stage++ {
		total, fail := 0, 0

		for _, playerStage := range stages {
			if playerStage >= stage {
				total += 1
				if playerStage == stage {
					fail += 1
				}
			}
		}

		failRate := 0.0
		if total > 0 {
			failRate = float64(fail) / float64(total)
		}

		failRates[stage-1] = Res2{stage, failRate}
	}

	sort.Slice(failRates, func(i, j int) bool {
		return failRates[i].failRate > failRates[j].failRate ||
			(failRates[i].failRate == failRates[j].failRate && failRates[i].stage < failRates[j].stage)
	})

	result := make([]int, N)
	for i, r := range failRates {
		result[i] = r.stage
	}

	return result
}

func main() {
	r1 := solution(5, []int{2, 1, 2, 6, 2, 4, 3, 3})
	fmt.Println(r1)

	r2 := solution(4, []int{4, 4, 4, 4, 4})
	fmt.Println(r2)

	r3 := solution(5, []int{2, 1, 2, 6, 2, 4, 3, 3})
	fmt.Println(r3)

	r4 := solution(4, []int{4, 4, 4, 4, 4})
	fmt.Println(r4)
}

File: src/olcw78 (die yoong)/jsAlgorithmStudy/week02/06_실패율.js
/**
 * 풀이
 * yet: 스테이지에 도달했으나 아직 클리어하지 못한 플레이어의 수
 * reached: 스테이지에 도달한 플레이어의 수
 * 실패율 정의 : yet / reached
 * 주어진 stages 를 돌며
 *
 * 분석
 * 시간 복잡도: O(N^2)
 *   - 2중 루프: O(N^2)
 *   - filter: O(M^2)
 *   - 오름차순 정렬: O(L logL)
 *   - O(N^2 + M^2 + L logL) -> O(2N^2) -> O(N^2)
 * 공간 복잡도: O(N)
 *   주어진 stages: number[] + 지울 대상 배열 + res 배열 -> O(3N) --> O(N)
 *
 * @param N {number} 전체 스테이지 개수
 * @param stages {number[]} 사용자가 현재 멈춰 있는 스테이지의 번호가 담긴 배열
 */
function solution(N, stages) {
    const res = []
    for (let i = 1; i <= N; i++) {
        let yet = 0;
        let reached = 0;

        const del = []
        for (const [sti, st] of stages.entries()) {
            reached++;

            if (st <= i) {
                yet++;
                del.push(sti);
            }
        }

        stages = stages.filter((_, i) => !del.includes(i));

        const failRate = yet / reached
        res.push({i, failRate})
    }

    res.sort((a, b) => b.failRate - a.failRate)
    return res.map(x => x.i)
}

const r1 = solution(5, [2, 1, 2, 6, 2, 4, 3, 3])
console.log(r1)

const r2 = solution(4, [4, 4, 4, 4, 4])
console.log(r2)

/**
 *
 * @param N {number}
 * @param stages {number[]}
 */
function answer(N, stages) {
    const challegener = new Array(N + 2).fill(0)
    for (const st of stages) {
        challegener[st] += 1
    }

    const fails = {}
    let total = stages.length
    for (let i = 0; i <= N; i++) {
        if (challegener[i] === 0) {
            fails[i] = 0
            continue
        }

        fails[i] = challegener[i] / total
        total -= challegener[i]
    }

    const res = Object.entries(fails)
        .sort((a, b) => b[1] - a[1])

    return res.map(x => Number(x[0]))
}